### PR TITLE
Adaptive Metropolis sampler (MCMC) 

### LINKED ISSUE #1397: [TASK] Adaptive MCMC Sampler for bayesian update
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**

Standard MCMC is not working well for high-dimension input space, while adaptive MCMC can be more efficient to generate the posterior distribution. 

**Describe the solution you'd like**
Implement adaptive MCMC algorithm. 

**Describe alternatives you've considered**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
close #1397 

##### What are the significant changes in functionality due to this change request?
Implementation of adaptive metropolis sampler for bayesian calibration. 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by wangcj05** [comment]
Can anyone @alfoa @mandd @joshua-cogliati-inl @PaulTalbot-INL take a look at this PR?

File: doc/user_manual/sampler.tex [hunk 1]
```diff
@@ -3285,3 +3294,130 @@ \subsubsection{Metropolis (Metropolis-Hastings Sampler)}
...
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%% Adaptive Metropolis Sampler %%%
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\subsubsection{Adaptive Metropolis Sampler}
+\label{subsubsubsec:adaptiveMetropolis}
+
+The search for improved proposal distributions of Metropolis sampler is often
+done manually, through trial and error, though this can be difficult especially
+in high dimensions. An alternative approach is adaptive Metropolis, which asks
+the computer to automatically "learn" better parameter values "on the fly".
```
Comment (alfoa) [inline doc/user_manual/sampler.tex:3307]: "learn" -> ``learn'' 

"on the fly" -> ``on the fly''

(the quotes to be more latex "friendly")

File: doc/user_manual/sampler.tex [hunk 2]
```diff
@@ -3285,3 +3294,130 @@ \subsubsection{Metropolis (Metropolis-Hastings Sampler)}
...
+      \item \xmlNode{initial}, \xmlDesc{float, optional field}, specified the initial value for given variable.
+      \item \xmlNode{proposal}, \xmlDesc{Assembler Object}, specifies the proposal distribution for this variable.
+      \nb We only allow one-dimensional symmetric distribution to be the proposal distribution.
+      This node must contain the following two attributes:
+      \begin{itemize}
+        \item \xmlAttr{class}, \xmlDesc{required string attribute}, the main
+          ``class'' of the listed object. Only ``Distributions'' is allowed.
+        \item \xmlAttr{type},  \xmlDesc{required string attribute}, the object
+          identifier or sub-type.
+        \item \xmlAttr{dim}, \xmlDesc{positive integer, optional attribute}, required for
```
Comment (alfoa) [inline doc/user_manual/sampler.tex:3335]: The proposal can  be multivariate or there is the same limitation in the ``not adaptive''?

File: doc/user_manual/sampler.tex [hunk 3]
```diff
@@ -3285,3 +3294,130 @@ \subsubsection{Metropolis (Metropolis-Hastings Sampler)}
...
+      \end{itemize}
+      \item \xmlNode{probabilityFunction}, \xmlDesc{Assembler Object}, specifies the prior distribution function..
+      This node must contain the following two attributes:
+      \begin{itemize}
+        \item \xmlAttr{class}, \xmlDesc{required string attribute}, the main
+          ``class'' of the listed object. Only ``Functions'' is allowed.
+        \item \xmlAttr{type},  \xmlDesc{required string attribute}, the object
+          identifier or sub-type. Only ``External'' is allowed.
+      \end{itemize}
+      \nb This node is currently not allowed for Adaptive Metropolis Sampler.
```
Comment (alfoa) [inline doc/user_manual/sampler.tex:3347]: can you capitalize and make this red? It is "difficult'' to see it when it is in the middle of the manual

File: framework/Samplers/MCMC/MCMC.py [hunk 4]
```diff
@@ -327,7 +357,61 @@ def localFinalizeActualSampling(self, jobObject, model, myInput):
...
+    ## tune scaling parameter
+    if not self._countsUntilTune and self._tune:
+      ### tune
+      self._scaling = self.tuneScalingParam(self._scaling, self._acceptInTune/float(self._tuneInterval))
+      ### reset counter
+      self._countsUntilTune = self._tuneInterval
+      self._acceptInTune = 0
+
+  @abc.abstractmethod
+  def _useRealization(self, newRlz, currentRlz):
```
Comment (alfoa) [inline framework/Samplers/MCMC/MCMC.py:395]: Do you think you need to "overload" this method to make this abstract  method? I see why you want to do it to force the derived classses to have this method but is it really required?

File: framework/Samplers/MCMC/MCMC.py [hunk 5]
```diff
@@ -367,3 +451,26 @@ def _addToSolutionExport(self, rlz):
     if self._burnIn < self.counter:
       rlz = dict((var, np.atleast_1d(val)) for var, val in rlz.items())
       self._solutionExport.addRealization(rlz)
+
+  @staticmethod
+  def tuneScalingParam(scale, acceptRate):
+    """
+      Tune the scaling parameter for the proposal distribution
+      Borrowed from PyMC3
```
Comment (alfoa) [inline framework/Samplers/MCMC/MCMC.py:459]: do we have a reference for these suggested scaling multiplier?

File: tests/framework/MCMC/test_1D.xml [hunk 6]
```diff
@@ -0,0 +1,102 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="silent">
+  <TestInfo>
+    <name>framework/MCMC.MetropolisOneDim</name>
+    <author>wangc</author>
+    <created>2020-12-20</created>
+    <classesTested>MCMC.Metropolis</classesTested>
+    <description>
+      Test the Markov Chain Monte Carlo algorithm: Metropolis Sampling
```
Comment (alfoa) [inline tests/framework/MCMC/test_1D.xml:9]: can you expand the description a bit more?

File: tests/framework/MCMC/test_10D_proposal_mvn.xml [hunk 7]
```diff
@@ -0,0 +1,186 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="silent">
```
Comment (alfoa) [inline tests/framework/MCMC/test_10D_proposal_mvn.xml:2]: should we create a new Requirement "satisfied" by these implementations? In my opinion we should.
Can you propose one?

**Review by alfoa — CHANGES_REQUESTED** [review]

**Comment by alfoa** [comment]
@wangcj05  few comments. Good job.

File: doc/user_manual/sampler.tex (hunk 1 above)
Comment (wangcj05) [reply doc/user_manual/sampler.tex]: fixed

File: doc/user_manual/sampler.tex (hunk 3 above)
Comment (wangcj05) [reply doc/user_manual/sampler.tex]: updated

File: framework/Samplers/MCMC/MCMC.py (hunk 5 above)
Comment (wangcj05) [reply framework/Samplers/MCMC/MCMC.py:459]: I do not find the reference. It seems to me the scaling is based on experience.

File: tests/framework/MCMC/test_1D.xml (hunk 6 above)
Comment (wangcj05) [reply tests/framework/MCMC/test_1D.xml:9]: updated

File: tests/framework/MCMC/test_10D_proposal_mvn.xml (hunk 7 above)
Comment (wangcj05) [reply tests/framework/MCMC/test_10D_proposal_mvn.xml:2]: Added

File: framework/Samplers/MCMC/MCMC.py (hunk 4 above)
Comment (wangcj05) [reply framework/Samplers/MCMC/MCMC.py:395]: I think it is needed.

**Comment by wangcj05** [comment]
@alfoa I have resolved your comments.

File: framework/Samplers/MCMC/MCMC.py (hunk 4 above)
Comment (alfoa) [reply framework/Samplers/MCMC/MCMC.py:395]: why?

**Review by alfoa — APPROVED** [review]

**Comment by alfoa** [comment]
Checklist passed...

Comments addressed...

PR can be merged...

**Pull request merged**